[SPARK-30144][ML][PySpark] Make MultilayerPerceptronClassificationModel extend MultilayerPerceptronParams

huaxingao opened the pull request:
### What changes were proposed in this pull request?
Make ```MultilayerPerceptronClassificationModel``` extend ```MultilayerPerceptronParams```


### Why are the changes needed?
Make ```MultilayerPerceptronClassificationModel``` extend ```MultilayerPerceptronParams``` to expose the training params, so user can see these params when calling ```extractParamMap```


### Does this PR introduce any user-facing change?
Yes. The ```MultilayerPerceptronParams``` such as ```seed```, ```maxIter``` ... are available in ```MultilayerPerceptronClassificationModel``` now


### How was this patch tested?
Manually tested ```MultilayerPerceptronClassificationModel.extractParamMap()``` to verify all the new params are there.

huaxingao committed:
[SPARK-30144][ML][PySpark] Make MultilayerPerceptronClassificationModel extend MultilayerPerceptronParams

SparkQA commented:
**[Test build #115115 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115115/testReport)** for PR 26838 at commit [`fc2cc5a`](https://github.com/apache/spark/commit/fc2cc5ab78c8fb7747c885c1583ac095fa68446a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19928/
Test PASSed.

SparkQA commented:
**[Test build #115115 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115115/testReport)** for PR 26838 at commit [`fc2cc5a`](https://github.com/apache/spark/commit/fc2cc5ab78c8fb7747c885c1583ac095fa68446a).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MultilayerPerceptronClassificationModel(JavaProbabilisticClassificationModel,`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115115/
Test FAILed.

huaxingao committed:
fix mima failure

SparkQA commented:
**[Test build #115121 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115121/testReport)** for PR 26838 at commit [`09bca1e`](https://github.com/apache/spark/commit/09bca1ea7d68ae161d7558a60bde2ac3e3ae51a8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19934/
Test PASSed.

SparkQA commented:
**[Test build #115121 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115121/testReport)** for PR 26838 at commit [`09bca1e`](https://github.com/apache/spark/commit/09bca1ea7d68ae161d7558a60bde2ac3e3ae51a8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115121/
Test PASSed.

dongjoon-hyun commented on project/MimaExcludes.scala:
```diff
@@ -328,6 +328,9 @@ object MimaExcludes {
     ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.status.api.v1.ApplicationEnvironmentInfo.this"),
 
+    // [SPARK-30144][ML] Make MultilayerPerceptronClassificationModel extend MultilayerPerceptronParams
+    ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.apache.spark.ml.classification.MultilayerPerceptronClassificationModel.layers"),
```
Just a question. Is this worth to break the API, @huaxingao ?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -273,29 +273,29 @@ object MultilayerPerceptronClassifier
  *
  * @param uid uid
- * @param layers array of layer sizes including input and output layers
+ * @param modelLayers array of layer sizes including input and output layers
```
I know the constructor is private, but is it necessary to change this name?

zhengruifeng replied:
I think it is needed, since all estimators and their models should share the same params.

zhengruifeng replied:
What about just removing `modelLayers` in model?
since the value (array of layer sizes) can be easily obtain by $(layers)

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -273,29 +273,29 @@ object MultilayerPerceptronClassifier
 class MultilayerPerceptronClassificationModel private[ml] (
     @Since("1.5.0") override val uid: String,
-    @Since("1.5.0") val layers: Array[Int],
+    @Since("1.5.0") val modelLayers: Array[Int],
```
same question, why need to change this?

viirya replied:
oh, i see. MultilayerPerceptronParams has layers too?

viirya replied:
Can we rename layers of MultilayerPerceptronParams instead?

zhengruifeng replied:
but renaming layers of MultilayerPerceptronParams will also break the API.

huaxingao commented:
@dongjoon-hyun @srowen @viirya Thanks for the review. 

Since ```MultilayerPerceptronParams``` has ```layers```, after ```MultilayerPerceptronClassificationModel``` extends ```MultilayerPerceptronParams```, I have to rename ```layers```.  It's not good to rename ```layers``` in  ```MultilayerPerceptronParams``` because  the getter/setter are public APIs.

```MultilayerPerceptronClassificationModel``` is the only one that doesn't have the training params. All the other XXXModel extend the corresponding XXXParams. In addition, as what is said in the description of the jira https://issues.apache.org/jira/browse/SPARK-30144, user wants to have a way to track what parameters are best during a crossvalidation, so I think it makes sense to expose  ```MultilayerPerceptronParams``` to ```MultilayerPerceptronClassificationModel```

huaxingao commented:
cc @zhengruifeng

zhengruifeng commented:
It seems that this will be a breaking change, so maybe we need to depreicate model's `layers` in 2.4.x?

zhengruifeng commented:
current var `layers` in `MultilayerPerceptronClassificationModel` is just `$(layers)` in the estimator, so I suggest just remove it in the model, and ignore it in model load.

huaxingao commented:
I can remove ```layers``` from ```MultilayerPerceptronClassificationModel```, but seems it may need more changes than current fix:
member variables ```numFeatures``` and ```mlpModel``` depend on ```layers``` at class initiation time, so these two need to be changed.
since no ```layers``` any more, writer and reader need to be changed accordingly.

So I guess probably keep the current fix?

srowen commented:
I see, so it's not just a question of the API, but the name of the field in the serialized model? Hm, yeah. I think it's OK to change and standardize the name in 3.0, but needs to be able to read 'layers' from previous models if present. Is that the only issue here?

zhengruifeng commented:
@huaxingao  I guess we can mark `numFeatures` & `mlpModel` lazy. Moreover, I think we should mark `mlpModel`   transient, since it is needed only in `transform`.

huaxingao committed:
address comments

SparkQA commented:
**[Test build #115309 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115309/testReport)** for PR 26838 at commit [`14ce378`](https://github.com/apache/spark/commit/14ce3781f922476c4cc7f863855544bc3f68ed68).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20117/
Test PASSed.

SparkQA commented:
**[Test build #115309 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115309/testReport)** for PR 26838 at commit [`14ce378`](https://github.com/apache/spark/commit/14ce3781f922476c4cc7f863855544bc3f68ed68).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115309/
Test PASSed.

huaxingao commented:
I manually saved 2.4.4 model and loaded using 3.0.0. It worked OK.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -369,11 +377,18 @@ object MultilayerPerceptronClassificationModel
+      } else {
+        val data = sparkSession.read.parquet(dataPath).select("weights").head()
+        val weights = data.getAs[Vector](0)
+        new MultilayerPerceptronClassificationModel(metadata.uid, weights)
```
Pardon the dumb question, but, why does the new model version not need to load layers? is it part of weights?

huaxingao replied:
In the new code, since ```MultilayerPerceptronClassificationModel``` extends ```MultilayerPerceptronParams```, ```layers``` is one of the Params, and it is set in the next line:
```
metadata.getAndSetParams(model)
```

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -273,29 +274,35 @@ object MultilayerPerceptronClassifier
     @Since("2.0.0") val weights: Vector)
   extends ProbabilisticClassificationModel[Vector, MultilayerPerceptronClassificationModel]
-  with Serializable with MLWritable {
+  with MultilayerPerceptronParams with Serializable with MLWritable {
```
Not related to this change. But do we use MultilayerPerceptronClassificationModel in executors? Like not every classification model extends Serializable.

huaxingao replied:
I am not sure about this. Seems only the tree related model extends Serializable.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -347,13 +355,13 @@ object MultilayerPerceptronClassificationModel
     override protected def saveImpl(path: String): Unit = {
       // Save metadata and Params
       DefaultParamsWriter.saveMetadata(instance, path, sc)
       // Save model data: layers, weights
```
no layers now.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -369,11 +377,18 @@ object MultilayerPerceptronClassificationModel
-      val model = new MultilayerPerceptronClassificationModel(metadata.uid, layers, weights)
-
+      val columns = sparkSession.read.parquet(dataPath).columns
+      val model = if (columns.length == 2) { // model prior to 3.0.0
```
Should we have an example old model and read it in one test case?

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -369,11 +377,18 @@ object MultilayerPerceptronClassificationModel
-      val weights = data.getAs[Vector](1)
-      val model = new MultilayerPerceptronClassificationModel(metadata.uid, layers, weights)
-
+      val columns = sparkSession.read.parquet(dataPath).columns
```
Hmm, I think we can read from the data path in any case. Then we check the length of returned array to decide we should read layers + weights or only weights. We do not need to read the data twice.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -273,29 +274,35 @@ object MultilayerPerceptronClassifier
 @Since("1.5.0")
 class MultilayerPerceptronClassificationModel private[ml] (
     @Since("1.5.0") override val uid: String,
-    @Since("1.5.0") val layers: Array[Int],
```
Shall we update migration guild?

huaxingao replied:
@srowen Sean, this question is for you.

huaxingao committed:
address comments

SparkQA commented:
**[Test build #115351 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115351/testReport)** for PR 26838 at commit [`7590bf8`](https://github.com/apache/spark/commit/7590bf85a30865c7f83914eb8b6f8b78589fd709).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20158/
Test PASSed.

SparkQA commented:
**[Test build #115351 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115351/testReport)** for PR 26838 at commit [`7590bf8`](https://github.com/apache/spark/commit/7590bf85a30865c7f83914eb8b6f8b78589fd709).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115351/
Test FAILed.

huaxingao commented:
retest this please

SparkQA commented:
**[Test build #115363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115363/testReport)** for PR 26838 at commit [`7590bf8`](https://github.com/apache/spark/commit/7590bf85a30865c7f83914eb8b6f8b78589fd709).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20170/
Test PASSed.

SparkQA commented:
**[Test build #115363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115363/testReport)** for PR 26838 at commit [`7590bf8`](https://github.com/apache/spark/commit/7590bf85a30865c7f83914eb8b6f8b78589fd709).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115363/
Test FAILed.

huaxingao committed:
fix test

SparkQA commented:
**[Test build #115374 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115374/testReport)** for PR 26838 at commit [`f98de6b`](https://github.com/apache/spark/commit/f98de6bf7cdbb31f7fc402b8010f63a10188d535).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20180/
Test PASSed.

zhengruifeng commented on mllib/src/test/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifierSuite.scala:
```diff
@@ -229,4 +229,17 @@ class MultilayerPerceptronClassifierSuite extends MLTest with DefaultReadWriteTe
       }
   }
+
+  test("Load MultilayerPerceptronClassificationModel prior to Spark 3.0") {
```
Just a general question, we also have other algorithms that modified the load/save method (like NaiveBayes), do we need to add testsuites for them like this?

huaxingao replied:
I am ok either way.

huaxingao replied:
so add similar test in ```NaiveBayes``` (and other algorithms that modified the load/save method) too?

zhengruifeng replied:
Maybe needed, but in other PRs.

SparkQA commented:
**[Test build #115374 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115374/testReport)** for PR 26838 at commit [`f98de6b`](https://github.com/apache/spark/commit/f98de6bf7cdbb31f7fc402b8010f63a10188d535).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115374/
Test FAILed.

viirya commented on python/pyspark/ml/classification.py:
```diff
@@ -2274,21 +2276,21 @@ def setSolver(self, value):
-    @since("1.6.0")
-    def layers(self):
+    @since("3.0.0")
+    def setLayers(self, value):
```
Do we need to add this setter? Do we allow to change layers of a model?

zhengruifeng replied:
I guess this method is for old version. Since the new version return layers via $(layers), so we need to manually set this param in old version.

viirya replied:
When we need to manually set layers param? I think it is fixed after training?

And we don't have other setters like setMaxIter.

huaxingao replied:
We manually set layers in loading the old model in scala. I think we normally keep the same methods between scala and python, so I put this in python too.

huaxingao replied:
Agree. I will use  ```model.set("layers", layers)``` in load. Will remove ```setLayers```.

CyborgDroid replied:
I opened the jira ticket and want to thank everyone for their work and comments. @viirya - I'm not sure I understood your question about setters, however, from a user perspective (me) the important thing is to expose what parameters were used to train the model. After running a cross validation, there is no other way of accessing parameters to analyze the results of the grid search. There is also no built in way to log the parameters during each grid search attempt.

Here you can see an example of how that is useful to narrow down the hyperparameter space based on which hyperparameters generally give good results.

https://github.com/CyborgDroid/spark_mlflow/blob/develop/README.md#mlflow-to-analyze-best-hyperparameters-for-further-optimization

viirya replied:
You can get the parameters used in training by getters. Getters are exposed.

CyborgDroid replied:
The purpose of the extractParamMap function is to expose all parameters used to train the model. Having to execute all getters is not a good approach or solution.

viirya replied:
Why extractParamMap does not work without setters?

viirya commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -247,7 +247,8 @@ class MultilayerPerceptronClassifier @Since("1.5.0") (
     val mlpModel = trainer.train(data)
-    new MultilayerPerceptronClassificationModel(uid, myLayers, mlpModel.weights)
+    val model = new MultilayerPerceptronClassificationModel(uid, mlpModel.weights)
+    model.setLayers(myLayers)
```
hmm, do we really have exposed parameters into the model? The purpose of this PR is to have training parameters in the model, but except for layers, seems other parameters are not set?

huaxingao replied:
The params are set at line 152 in Preditor

```copyValues(train(casted).setParent(this))``` 

It copies all the params in ```paramMap``` from ```Preditor``` (```MultilayerPerceptronClassifier```) to the model. Actually since ```layers``` is one of the params now, it also sets here. No need to set it explicitly. I will remove line 251.

huaxingao committed:
fix test failure

huaxingao committed:
remove obsolete files

SparkQA commented:
**[Test build #115403 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115403/testReport)** for PR 26838 at commit [`6be731d`](https://github.com/apache/spark/commit/6be731dd7ab723f925a6933ce7d2f252fe3b1fc0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20208/
Test PASSed.

SparkQA commented:
**[Test build #115403 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115403/testReport)** for PR 26838 at commit [`6be731d`](https://github.com/apache/spark/commit/6be731dd7ab723f925a6933ce7d2f252fe3b1fc0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115403/
Test FAILed.

viirya commented:
retest this please.

SparkQA commented:
**[Test build #115413 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115413/testReport)** for PR 26838 at commit [`6be731d`](https://github.com/apache/spark/commit/6be731dd7ab723f925a6933ce7d2f252fe3b1fc0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20218/
Test PASSed.

SparkQA commented:
**[Test build #115413 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115413/testReport)** for PR 26838 at commit [`6be731d`](https://github.com/apache/spark/commit/6be731dd7ab723f925a6933ce7d2f252fe3b1fc0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115413/
Test PASSed.

huaxingao committed:
remove setLayers

SparkQA commented:
**[Test build #115423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115423/testReport)** for PR 26838 at commit [`7a98ffb`](https://github.com/apache/spark/commit/7a98ffbc780770bfb6454ea72a597b1c0fb168d1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20226/
Test PASSed.

SparkQA commented:
**[Test build #115423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115423/testReport)** for PR 26838 at commit [`7a98ffb`](https://github.com/apache/spark/commit/7a98ffbc780770bfb6454ea72a597b1c0fb168d1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115423/
Test FAILed.

huaxingao commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20232/
Test PASSed.

SparkQA commented:
**[Test build #115429 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115429/testReport)** for PR 26838 at commit [`7a98ffb`](https://github.com/apache/spark/commit/7a98ffbc780770bfb6454ea72a597b1c0fb168d1).

SparkQA commented:
**[Test build #115429 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115429/testReport)** for PR 26838 at commit [`7a98ffb`](https://github.com/apache/spark/commit/7a98ffbc780770bfb6454ea72a597b1c0fb168d1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115429/
Test FAILed.

zhengruifeng commented:
retest this please

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -369,11 +361,18 @@ object MultilayerPerceptronClassificationModel
-      val model = new MultilayerPerceptronClassificationModel(metadata.uid, layers, weights)
-
+      val df = sparkSession.read.parquet(dataPath)
+      val model = if (df.columns.length == 2) { // model prior to 3.0.0
```
totally nit: why not using version info in var `metadata` like other algs?
I guess we can fetch the main version from it even if the params are not saved in it in the old version.

huaxingao replied:
I will fix this. Thanks.

SparkQA commented:
**[Test build #115443 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115443/testReport)** for PR 26838 at commit [`7a98ffb`](https://github.com/apache/spark/commit/7a98ffbc780770bfb6454ea72a597b1c0fb168d1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20246/
Test PASSed.

zhengruifeng commented:
I just notice that the previous MLP mode is saved in  `mllib/src/test/resources/test-data/mlp-pre3.0/`
1, what about rename the dir like `ml-models`? the `test-data` is more like for a dataset;
2, what version do you use to generate the model? pre-3.0? it maybe not a good choice, I think we may use a public released version like 2.4.4

SparkQA commented:
**[Test build #115443 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115443/testReport)** for PR 26838 at commit [`7a98ffb`](https://github.com/apache/spark/commit/7a98ffbc780770bfb6454ea72a597b1c0fb168d1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115443/
Test PASSed.

huaxingao committed:
address comments

huaxingao commented:
@zhengruifeng Sorry for the late reply. Was extremely busy with something in the past few days.

I renamed the dir to ```ml-models``` per your suggestion. Since we have old models for stringIndexer and HashingTF in that dir too, I also renamed the path in ```StringIndexerSuite``` and ```HashingTFSuite```

I used 2.4.4 to generate the old model.

viirya commented:
The change looks good now. It could be a good time at 3.0.0 to do such change, though we do not deprecate layers in 2.4. Not sure if it is a problem.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20387/
Test PASSed.

SparkQA commented:
**[Test build #115587 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115587/testReport)** for PR 26838 at commit [`fdfeb6b`](https://github.com/apache/spark/commit/fdfeb6b94bb8de2849ca33d2fba50d0107f7dbfa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115587/
Test PASSed.

zhengruifeng commented:
LGTM

srowen reviewed (commented):
OK so the only remaining question is what to do about the breaking API change here. Can we leave in existing accessors as aliases as much as possible, and deprecate them? then there's no real issue here

huaxingao commented:
So I need to deprecate ```layers``` in 2.4.5, right?

srowen commented:
Can we just keep the existing layers getters as an alias, but make them deprecated, for 3.0?

huaxingao commented:
@srowen Do you mean something like this:
```
class MultilayerPerceptronClassificationModel private[ml] (
    @Since("1.5.0") override val uid: String,
    @Since("1.5.0")
    @deprecated("This is deprecated and will be removed in the future", "3.0.0")
    val modelLayers: Array[Int],
    @Since("2.0.0") val weights: Vector)
```
I have to change to ```modelLayers``` because ```layers``` conflicts with ```layers``` param inherited from ```MultilayerPerceptronParams```.

srowen commented:
I really meant the various `def layers` that are removed. They could be kept and deprecated, and avoid MiMa exclusions, I think. Yes you have to change the name in the constructors, but those are private anyway. I'm just trying to avoid breaking changes.

huaxingao commented:
The Mima exclusion 

```ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.apache.spark.ml.classification.MultilayerPerceptronClassificationModel.layers")``` 

is because of the name collision between the original constructor parameter 
```val layers: Array[Int]``` and the ```layers``` param inherited from MultilayerPerceptronParams.

```
[error]  * method layers()Array[Int] in class org.apache.spark.ml.classification.MultilayerPerceptronClassificationModel has a different result type in current version, where it is org.apache.spark.ml.param.IntArrayParam rather than Array[Int]
```

It's not because ```def layers``` is removed. So it seems to me this Mima exclusion can't be avoided.

srowen commented:
OK. I wonder if it's worthwhile keeping the existing getters anyway -- the ones that are explicitly removed here -- and deprecating them, even if we'd still have a breaking change and MiMa exclusions. 

I think the change is OK if we can tag the JIRA with release-notes and add a note explaining the need to migrate to a new field name.

huaxingao committed:
address comments

SparkQA commented:
**[Test build #115826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115826/testReport)** for PR 26838 at commit [`94b51a7`](https://github.com/apache/spark/commit/94b51a7f4734222d1c8a46ac7819650024b60eb8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20618/
Test PASSed.

SparkQA commented:
**[Test build #115826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115826/testReport)** for PR 26838 at commit [`94b51a7`](https://github.com/apache/spark/commit/94b51a7f4734222d1c8a46ac7819650024b60eb8).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115826/
Test FAILed.

huaxingao commented:
I tagged the JIRA with release_notes and added a note. 
https://issues.apache.org/jira/browse/SPARK-30144

I also put back the ```layers``` in python, but I am thinking of removing it again: In the note, I have explained that ```layers``` is changed from Array[Int] to IntArrayParam so ```getLayers``` should be used instead of ```layers```. Seems to me, there is no need to keep ```layers```.

srowen commented:
I see, so you mean there is no meaningful way to keep the previous setters, because the nature of the param has changed?

huaxingao commented:
Yes.

srowen commented:
OK, if this is the least change we can make while fixing the inconsistency, I'd be OK with it for 3.0. Even though these weren't technically deprecated, it's a minor API and a minor change, and still legitimate for a major release, versus keeping the inconsistency for years.

huaxingao committed:
remove layers

SparkQA commented:
**[Test build #115871 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115871/testReport)** for PR 26838 at commit [`1833754`](https://github.com/apache/spark/commit/18337544c09bf1bfc57218d20e43db5094b07e41).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20663/
Test PASSed.

SparkQA commented:
**[Test build #115871 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115871/testReport)** for PR 26838 at commit [`1833754`](https://github.com/apache/spark/commit/18337544c09bf1bfc57218d20e43db5094b07e41).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115871/
Test FAILed.

huaxingao commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20664/
Test PASSed.

SparkQA commented:
**[Test build #115872 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115872/testReport)** for PR 26838 at commit [`1833754`](https://github.com/apache/spark/commit/18337544c09bf1bfc57218d20e43db5094b07e41).

SparkQA commented:
**[Test build #115872 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115872/testReport)** for PR 26838 at commit [`1833754`](https://github.com/apache/spark/commit/18337544c09bf1bfc57218d20e43db5094b07e41).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115872/
Test PASSed.

zhengruifeng commented:
@huaxingao It seems that the models are saved in different dir?

huaxingao commented:
@zhengruifeng Yes. The models are now in mllib/src/test/resources/ml-models

zhengruifeng commented:
@huaxingao It seems that hashingTF/strIndexer models are stored in `test-data`.
Totally nit, I think it is better to rename them `-2.4.4` instead of `-pre3.0`, otherwise we may think that they are generated by version `Preview release of Spark 3.0`

huaxingao committed:
change dir from mlp-pre3.0 to mlp-2.4.4

SparkQA commented:
**[Test build #115969 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115969/testReport)** for PR 26838 at commit [`07267ff`](https://github.com/apache/spark/commit/07267ff59a127e01783fdd4964583023f98aba55).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20761/
Test PASSed.

huaxingao commented:
@zhengruifeng 
I changed the dir to mlp-2.4.4. I checked hashingTF/strIndexer models, they are in ml-models. 

![image](https://user-images.githubusercontent.com/13592258/71597172-0fe25000-2af7-11ea-8b69-78812f950280.png)

SparkQA commented:
**[Test build #115969 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115969/testReport)** for PR 26838 at commit [`07267ff`](https://github.com/apache/spark/commit/07267ff59a127e01783fdd4964583023f98aba55).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115969/
Test FAILed.

huaxingao committed:
fix test failure

SparkQA commented:
**[Test build #115971 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115971/testReport)** for PR 26838 at commit [`fa1797e`](https://github.com/apache/spark/commit/fa1797e8904bca380967beac7080562bcc2e64f8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20763/
Test PASSed.

SparkQA commented:
**[Test build #115971 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115971/testReport)** for PR 26838 at commit [`fa1797e`](https://github.com/apache/spark/commit/fa1797e8904bca380967beac7080562bcc2e64f8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115971/
Test PASSed.

zhengruifeng commented on mllib/src/test/scala/org/apache/spark/ml/feature/StringIndexerSuite.scala:
```diff
@@ -459,7 +459,7 @@ class StringIndexerSuite extends MLTest with DefaultReadWriteTest {
   }
 
   test("Load StringIndexderModel prior to Spark 3.0") {
-    val modelPath = testFile("test-data/strIndexerModel")
```
strIndexerModel-2.4.4？

zhengruifeng commented on mllib/src/test/scala/org/apache/spark/ml/feature/HashingTFSuite.scala:
```diff
@@ -89,7 +89,7 @@ class HashingTFSuite extends MLTest with DefaultReadWriteTest {
 
   test("SPARK-23469: Load HashingTF prior to Spark 3.0") {
-    val hashingTFPath = testFile("test-data/hashingTF-pre3.0")
+    val hashingTFPath = testFile("ml-models/hashingTF-pre3.0")
```
hashingTF-2.4.4

huaxingao replied:
updated. Thanks!

huaxingao committed:
change model path to 2.4.4

SparkQA commented:
**[Test build #115997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115997/testReport)** for PR 26838 at commit [`40fc5da`](https://github.com/apache/spark/commit/40fc5da2c741af0a806fa9fecc9b52ecddc9e0bb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20789/
Test PASSed.

SparkQA commented:
**[Test build #115997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115997/testReport)** for PR 26838 at commit [`40fc5da`](https://github.com/apache/spark/commit/40fc5da2c741af0a806fa9fecc9b52ecddc9e0bb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115997/
Test PASSed.

srowen commented:
Merged to master

huaxingao commented:
Thank you all!
